### PR TITLE
Switch accounts storage lock to DashMap

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Accounts scans from RPC hold:

1) the account storage lock duration the entire duration of the scan, which blocks replay on `create_and_insert_store()` during account commit.

2)  the account index lock duration the entire duration of the scan, which blocks replay in AccountsDb `store()->update_index()` during account commit. 

#### Summary of Changes
Experimenting with switching global accounts storage (part 1 above) lock to `DashMap`: https://github.com/xacrimon/dashmap, a concurrent hashmap implemented by sharding the table. This removes the need to hold the global `AccountStorage` read lock in `scan_accounts` which is blocking `create_and_insert_store()`

TODO: This can also potentially be expanded to replace the accounts index lock in part 2 above

TODO: Reason about correctness of places where I've replaced the large account_storage read locks, specifically around cleaning and shrinking. @ryoqun would really appreciate a review in those areas!

Pertinent gotchas with v3 of `DashMap`: https://github.com/xacrimon/dashmap/issues/74

Other candidates that were considered but not chosen:

1) chashmap: No iterator support
2) evmap: Poorer write performance under heavy read contention
3) flurry (Java concurrent hashmap port): Not as performant as DashMap

Fixes #

### CODE REVIEW

**Comment by t-nelson** [comment]
> Pertinent gotchas with v3 of DashMap: xacrimon/dashmap#74

The 4.0.0 release candidates are alleged to resolve this issue

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -702,10 +707,9 @@ impl AccountsDB {
         // Calculate store counts as if everything was purged
         // Then purge if we can
         let mut store_counts: HashMap<AppendVecId, (usize, HashSet<Pubkey>)> = HashMap::new();
-        let storage = self.storage.read().unwrap();
```
Comment (carllin) [inline runtime/src/accounts_db.rs:705]: From my understanding, the only extra risk this adds here, is that where before holding this lock would prevent any other slots in `self.storage` from being modified, now other slots can be modified. The three places this can happen

1. https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L970-L974. Shouldn't be possible b/c we hold the `shrink_candidate_slots` lock, and `shrink_all_slots() -> do_shrink_slot_forced()` is not called after startup, so the lock should be respected  

2. https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1256-L1259. Can this remove a slot that exists in `account_infos` such that the call below to `self.storage.0.get(&slot).unwrap();` panics?

3. https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1236-L1238. Adding a new storage entry should be ok as that should be on some future non-rooted slot which shouldn't exist in `account_infos`

**Comment by ryoqun** [comment]
> TODO: This can also potentially be expanded to replace the accounts index lock in part 2 above

Btw, this needs concurrent btree map because currently AccountsIndex uses `BTreeMap`  for predictable rent collection scanning.
Maybe like https://docs.rs/concread/0.2.3/concread/bptree/struct.BptreeMap.html

File: runtime/src/accounts_db.rs [hunk 2]
```diff
@@ -381,7 +386,7 @@ pub struct AccountsDB {
     /// Keeps tracks of index into AppendVec on a per slot basis
     pub accounts_index: RwLock<AccountsIndex<AccountInfo>>,
 
-    pub storage: RwLock<AccountStorage>,
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:384]: cool :)

File: runtime/src/accounts_db.rs [hunk 3]
```diff
@@ -1232,9 +1220,7 @@ impl AccountsDB {
         let store =
             Arc::new(self.new_storage_entry(slot, &Path::new(&self.paths[path_index]), size));
         let store_for_index = store.clone();
-
-        let mut stores = self.storage.write().unwrap();
-        let slot_storage = stores.0.entry(slot).or_insert_with(HashMap::new);
+        let mut slot_storage = self.storage.0.entry(slot).or_insert_with(HashMap::new);
         slot_storage.insert(store.id, store_for_index);
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1224]: yay! :)

File: runtime/src/accounts_db.rs [hunk 4]
```diff
@@ -1178,15 +1168,13 @@ impl AccountsDB {
...
-        Self::load(&storage, ancestors, &accounts_index, pubkey)
+        Self::load(&self.storage, ancestors, &accounts_index, pubkey)
     }
 
     fn find_storage_candidate(&self, slot: Slot) -> Arc<AccountStorageEntry> {
         let mut create_extra = false;
-        let stores = self.storage.read().unwrap();
-
-        if let Some(slot_stores) = stores.0.get(&slot) {
+        let slot_stores_guard = self.storage.0.get(&slot);
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1176]: nits: I think this isn't a lock guard anymore. So, remove the `_guard` prefix?

File: runtime/src/accounts_db.rs (hunk 1 above)
Comment (ryoqun) [reply runtime/src/accounts_db.rs:705]: @carllin 

1. Yeah, I think your understanding is correct.
2. Yeah, can `panic!`. you are addressing this at #12194 
3. Yeah, this is correct as well.

**Comment by ryoqun** [comment]
I think this is better than #12132 because of its straight-forwardness to the problem.

Anyway, I wonder how to best to fix the locks for AccountsDB...

File: runtime/src/accounts_db.rs (hunk 1 above)
Comment (carllin) [reply runtime/src/accounts_db.rs:705]: Awesome, thanks!

**Comment by carllin** [comment]
@ryoqun I think this PR and  #12132 may actually work together.

DashMap essentially partitions the HashMap into shards based on the value hash so there's not one giant lock, but internally, values with the same hash may still contend with each other.

#12132 should guarantee that during the accounts scan, a single shard isn't locked up for the entire duration of the scan within that shard

File: runtime/src/accounts_db.rs (hunk 4 above)
Comment (carllin) [reply runtime/src/accounts_db.rs:1176]: agreed

**Comment by carllin** [comment]
> > TODO: This can also potentially be expanded to replace the accounts index lock in part 2 above
> 
> Btw, this needs concurrent btree map because currently AccountsIndex uses `BTreeMap` for predictable rent collection scanning.
> Maybe like https://docs.rs/concread/0.2.3/concread/bptree/struct.BptreeMap.html

Unfortunately that one doesn't seem to implement the "range()" function we want from BTreeMap :cry: , thought it may be an easy addition.

I think a sufficient temporary bandaid is to switch `AccountsIndex.account_maps` to a `pub account_maps: AccountMap<Pubkey, Arc<AccountMapEntry<T>>>,` and clone out batches of the the `Arc` during the scan. It'll slow down the iteration (we can use a bigger batching factor), but at least it won't block the lock

File: runtime/src/accounts_db.rs (hunk 4 above)
Comment (carllin) [reply runtime/src/accounts_db.rs:1176]: heh well technically, I think it's still a guard on the specific DashMap shard: https://docs.rs/dashmap/3.11.10/src/dashmap/lib.rs.html#432-438, i.e. you shouldn't hold it for too long.

Comment (ryoqun) [reply runtime/src/accounts_db.rs:1180]: I see!

**Comment by ryoqun** [comment]
> Unfortunately that one doesn't seem to implement the "range()" function we want from BTreeMap cry , thought it may be an easy addition.

Sad...

> I think a sufficient temporary bandaid is to switch `AccountsIndex.account_maps` to a `pub account_maps: AccountMap<Pubkey, Arc<AccountMapEntry<T>>>,` and clone out batches of the the `Arc` during the scan. It'll slow down the iteration (we can use a bigger batching factor), but at least it won't block the lock

Yeah, that will work. Seems a good bandaid. :)

**Comment by ryoqun** [comment]
@carllin how does this perform in various benchmarks? There is several benchmark standard problems. Also, could you try to simulate concurrent heavy write by ReplayStage  and heavy read by RPC?

**Comment by carllin** [comment]
@ryoqun I added a benchmark simulating heavy single reads from RPC along with writes from Replay.

As expected, this case doesn't see a lot of benefit from this change because:

1) Most of the time `store` is blocked on the AccountsIndex lock. This is especially bad on Linux as the RwLock is read-favored so the read locks starve the write lock. 

2) `store` only needs the account storage write lock a couple of times to create the AccountStorageEntry. Afterwards, it just uses a read lock.

I didn't yet add a benchmark here for simulating the RPC `scan` case which I expect to see the most improvement (that benchmark should be an easy extension of this one). I'll add that in with the AccountsIndex change.

File: runtime/src/accounts_db.rs [hunk 5]
```diff
@@ -121,31 +122,31 @@ impl Versioned for (u64, AccountInfo) {
 }
 
 #[derive(Clone, Default, Debug)]
-pub struct AccountStorage(pub HashMap<Slot, SlotStores>);
+pub struct AccountStorage(pub DashMap<Slot, Arc<RwLock<SlotStores>>>);
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:125]: nits: how about adjusting the type alias `SlotStores` and hide `Arc<RwLock<_>>` here?

File: runtime/src/accounts_db.rs [hunk 6]
```diff
@@ -121,31 +122,31 @@ impl Versioned for (u64, AccountInfo) {
...
-                        .get_account(account_info.offset)?
-                        .0
-                        .clone_account(),
-                )
-            })
-            .map(|account| (account, slot))
+            .and_then(|storage_map| storage_map.value().read().unwrap().get(&store_id).cloned())
+    }
+
+    fn get_slot_stores(&self, slot: Slot) -> Option<Arc<RwLock<SlotStores>>> {
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:138]: nits: how about adjusting the type alias `SlotStores` and hide `Arc<RwLock<_>>` here?

File: runtime/src/accounts_db.rs [hunk 7]
```diff
@@ -1241,14 +1246,14 @@ impl AccountsDB {
...
-                    .scan_accounts(account_info, slot)
-                    .map(|(account, slot)| (pubkey, account, slot)),
-            )
+            let account_storage_entry = self
+                .storage
+                .get_account_storage_entry(slot, account_info.store_id);
+            let account_slot = account_storage_entry
+                .and_then(|account_storage_entry| account_storage_entry.get_account(account_info))
+                .map(|account| (pubkey, account, slot));
+            scan_func(&mut collector, account_slot)
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1256]: this

File: runtime/src/accounts_db.rs [hunk 8]
```diff
@@ -1221,14 +1226,14 @@ impl AccountsDB {
...
-                storage
-                    .scan_accounts(account_info, slot)
-                    .map(|(account, slot)| (pubkey, account, slot)),
-            )
+            let account_storage_entry = self
+                .storage
+                .get_account_storage_entry(slot, account_info.store_id);
+            let account_slot = account_storage_entry
+                .and_then(|account_storage_entry| account_storage_entry.get_account(account_info))
+                .map(|account| (pubkey, account, slot));
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1235]: nits: how about defining `AccountsDb::get_account_from_storage()` with these code and share with [that](https://github.com/solana-labs/solana/pull/12126/files#r503739183)?

File: runtime/src/accounts_db.rs [hunk 9]
```diff
@@ -2140,23 +2133,24 @@ impl AccountsDB {
...
-                    if count == 0 {
-                        dead_slots.insert(*slot);
-                    }
+            if let Some(store) = self
+                .storage
+                .get_account_storage_entry(*slot, account_info.store_id)
+            {
+                assert_eq!(
+                    *slot, store.slot,
+                    "AccountDB::accounts_index corrupted. Storage should only point to one slot"
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:2142]: nits: not this pr specific changes, but how about printing `*slot` in `store.slot` in the custom message? this would be crutial clue for debugging this, if this occurs ever after this pr...  ;)

File: runtime/src/accounts_db.rs [hunk 10]
```diff
@@ -2368,8 +2363,7 @@ impl AccountsDB {
 
     pub fn generate_index(&self) {
         let mut accounts_index = self.accounts_index.write().unwrap();
-        let storage = self.storage.read().unwrap();
-        let mut slots: Vec<Slot> = storage.0.keys().cloned().collect();
+        let mut slots: Vec<Slot> = self.storage.all_slots();
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:2366]: nits: no need to spell out `Vec<Slot>`?

File: runtime/src/accounts_db.rs [hunk 11]
```diff
@@ -1104,10 +1112,8 @@ impl AccountsDB {
             start.stop();
             handle_reclaims_elapsed = start.as_us();
 
-            let mut start = Measure::start("write_storage_elapsed");
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1107]: @carllin oops. maybe unintentional remove? We're starting mixing the measured time for `"write_storage_elapsed"` with `"update_index_elapsed"` by removing `"write_storage_elapsed"`.

File: programs/bpf/tests/programs.rs [hunk 12]
```diff
@@ -105,8 +105,13 @@ fn process_transaction_and_record_inner(
     let signature = tx.signatures.get(0).unwrap().clone();
     let txs = vec![tx];
     let tx_batch = bank.prepare_batch(&txs, None);
-    let (mut results, _, mut inner, _transaction_logs) =
-        bank.load_execute_and_commit_transactions(&tx_batch, MAX_PROCESSING_AGE, false, true, false);
+    let (mut results, _, mut inner, _transaction_logs) = bank.load_execute_and_commit_transactions(
```
Comment (ryoqun) [inline programs/bpf/tests/programs.rs:108]: nits: unrelated formatting change?

File: runtime/src/accounts_db.rs (hunk 11 above)
Comment (carllin) [reply runtime/src/accounts_db.rs:1107]: ah good catch, thanks!

File: runtime/src/accounts_db.rs (hunk 9 above)
Comment (carllin) [reply runtime/src/accounts_db.rs:2142]: yup, definitely would be helpful, added!

File: programs/bpf/tests/programs.rs (hunk 12 above)
Comment (carllin) [reply programs/bpf/tests/programs.rs:108]: reverted!

File: runtime/src/accounts_db.rs (hunk 8 above)
Comment (carllin) [reply runtime/src/accounts_db.rs:1235]: cleaned up!

File: runtime/src/accounts_db.rs [hunk 13]
```diff
@@ -1260,27 +1265,19 @@ impl AccountsDB {
         F: Fn(&StoredAccount, AppendVecId, &mut B) + Send + Sync,
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1260]: I think the above (out of the context diff) comemnt below isn't relevant or severely outdated anymore:

```
    // PERF: Sequentially read each storage entry in parallel
```

File: runtime/benches/accounts.rs [hunk 14]
```diff
@@ -141,3 +142,57 @@ fn bench_delete_dependencies(bencher: &mut Bencher) {
         accounts.accounts_db.clean_accounts(None);
     });
 }
+
+#[bench]
+#[ignore]
+fn bench_concurrent_read_write(bencher: &mut Bencher) {
```
Comment (ryoqun) [inline runtime/benches/accounts.rs:148]: btw, do you have any idea of how _single-threaded_ throughput of writing is changed from `std::collections::HashMap` to `DashMap`? I think I'm worrying too much but I'd rather want to confirm how far does `DashMap` do well while supporting the concurrency via sharding. Maybe it's trading off maximum throughput by negligible margin?

Comment (ryoqun) [reply runtime/benches/accounts.rs:148]: To add more context, our basic tenet is _batch them if we can, make it concurrent otherwise_. so, I guess the upper layers slumming the AccountsDB optimized for batching and the single threaded perf is kind of moderately related to the batching perf. Thus, we're somewhat sensitive to it.

**Comment by ryoqun** [comment]
> almost lgtm

special thanks for addressing my bunch of nits quickly as always. this pr got mature pretty quickly because of it. :)

File: runtime/benches/accounts.rs (hunk 14 above)
Comment (carllin) [reply runtime/benches/accounts.rs:148]: @ryoqun that's a good point, I've run the benchmark here: https://github.com/xacrimon/conc-map-bench which has three different work profiles, exchange(read-write), cache (read-heavy), and rapid-grow (insert-heavy), more details can be found in that link above. The results for a single thread on my Macbook Pro: 

== cache
-- MutexStd
25165824 operations across 1 thread(s) in 14.941146454s; time/op = 593ns

-- DashMap
25165824 operations across 1 thread(s) in 15.589596263s; time/op = 619ns
==

== exchange
-- MutexStd
25165824 operations across 1 thread(s) in 20.954264682s; time/op = 831ns

-- DashMap
25165824 operations across 1 thread(s) in 20.875345754s; time/op = 828ns
==

== rapid grow
-- MutexStd
25165824 operations across 1 thread(s) in 20.22593938s; time/op = 802ns
==

-- DashMap
25165824 operations across 1 thread(s) in 17.456807471s; time/op = 693ns
==

**Review by ryoqun — APPROVED** [review]
LGTM in code-wise with all nits resolved correctly!

Please check some perf. concerns I wrote about.

File: runtime/benches/accounts.rs (hunk 14 above)
Comment (carllin) [reply runtime/benches/accounts.rs:148]: @ryoqun, good suggestions, here's the results I saw on my Macbook Pro:

`bench_concurrent_read_write` on 1 writer, no readers:`

```
DashMap:
test bench_concurrent_read_write  ... bench:   3,713,260 ns/iter (+/- 679,081)

Master: 
test bench_concurrent_read_write  ... bench:   3,773,731 ns/iter (+/- 654,643)
```

`accounts-bench/src/main.rs:`

```
DashMap:
clean: false
Creating 10000 accounts
created 10000 accounts in 4 slots create accounts took 148ms

Master: 
clean: false
Creating 10000 accounts
created 10000 accounts in 4 slots create accounts took 145ms
```

Comment (ryoqun) [reply runtime/benches/accounts.rs:148]: @carllin Perfect reporting :)
